### ISSUE TITLE
BGP ASN Conflict

### ISSUE THREAD

**Issue by melusom** [awslabs/aws-transit-vpc#27]
Starting 6/30/2018, all VGWs have a [default ASN](https://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_VPN.html#VPN) of 64512. This project sets a [default value](https://github.com/awslabs/aws-transit-vpc/blob/master/deployment/transit-vpc-primary-account.template#L36) for the customer gateway as 64512. Therefore, anyone who launches this solution without changing the defaults will have errors in their Lambda execution:

```
An error occurred (AsnConflict) when calling the CreateVpnConnection operation: The ASN of the specified customer gateway and virtual private gateway are the same.: ClientError
Traceback (most recent call last):
File "/var/task/transit-vpc-poller.py", line 167, in lambda_handler
vpn1=ec2.create_vpn_connection(Type='ipsec.1',CustomerGatewayId=cg1['CustomerGateway']['CustomerGatewayId'],VpnGatewayId=vgw['VpnGatewayId'],Options={'StaticRoutesOnly':False})
File "/var/runtime/botocore/client.py", line 314, in _api_call
return self._make_api_call(operation_name, kwargs)
File "/var/runtime/botocore/client.py", line 612, in _make_api_call
raise error_class(parsed_response, operation_name)
ClientError: An error occurred (AsnConflict) when calling the CreateVpnConnection operation: The ASN of the specified customer gateway and virtual private gateway are the same.
```

Kudos to @grolston for helping diagnose the issue.

**Comment by grolston** [comment]
Credit to @kmoschkau as well. We are currently looking at changing the defaults though we are looking to see the implications on the current environment for already provisioned VPCs.  The error will only happen on any newly created VPCs and logged in the vgw poller lambda function.

**Comment by jofwar** [comment]
After changing the default ASN on my VGW, the VPN Connections are reporting IPSEC IS DOWN. Can't for the life of me figure out a fix...

anyone know a remedy for this?

**Comment by greg-egt-2017** [comment]
put it back to default ASN (ASN 7224), unless you have an explicit need for non-default ASN

**Comment by grolston** [comment]
We noticed today that the new vpc configs were not being pushed back to the central S3 bucket for the CSRs. We did not get any errors in the poller lambda function but the routes in the new VPC were not being propagated even though propagation was turned on. 

Right now we were able to get routing to work when we deployed a new VPC via CloudFormation and set the VGW ASN to 64513. Prior to deploying the CFN stack we did set our  vpn_config.txt file BGP ASN back to the previous value of 64512.

We are testing to see if all works, but we are able to setup new VPCs in old accounts that route successfully.

cc @kmoschkau

**Comment by grolston** [comment]
@jofwar

Did it break all transit-vpc routing? When you made the changes did you have any issues with the CSRs?

**Comment by Jason-Crowe** [comment]
we just hit this today also. we would like some advice regarding what changes we should make and whether or not we must rebuild our entire transit-vpc installation

**Comment by grolston** [comment]
We have a ticket open with AWS and scheduled some time to discuss with their support. We will post what we find here..

**Comment by dbektas** [comment]
Any updates on this issue or suggestions that wouldn't break existing routing?

**Comment by kmoschkau** [comment]
We had a long support call with AWS support yesterday on this. We found that it is impossible to update the transit VPC config to use a different ASN, as the CSRs cannot support BGP entries with different ASNs. It was recommended that we deploy new VGWs with an updated ASN. When we tried this, we ran into another problem specific to our configuration (maxed out the # of BGP entries in our route tables). So, we will be consolidating them first and continuing support with AWS to determine whether deploying new VGWs will work. They informed us that it has worked for other customers.

**Comment by kskkskksk** [comment]
It seems to have been discussing ASN, so I'd like to confirm it.
When you use Transit VPC via Direct Connect newly, can you communicate ASN in the following example?

Peering BGP : 65000
VGW (Not Attach VPC) : 64512
Transit VPC : 64612 ※ Conflict at 64512
VGW on Spoke VPC : 64512

**Comment by grolston** [comment]
What has worked for us was changing the VGW ASN in our CloudFormation to not use the default 64512 (in our case we simply used 64513). This wired up fine and only involved us changing our CloudFormation that deploys VPCs that leverage transit-vpc.

**Comment by grolston** [comment]
Regarding our issue @kmoschkau  defined we did have a workaround we can share:

Our next issue which is outside of the ASN assignment was that the Route Table limitation imposed by AWS was 100. What we ended up doing is pushing everything through transit VPC (even outbound to internet) so we simplified our route table to 0.0.0.0/0 to VGW. If we had any additional routes to be used we would add them to our CFN deployments, but this was FAR less than the propagated routes coming through (we went from 100 to 4 routes in total per route table). The next issue we had was the limitation on the BGP sessions allowed in the VGW tunnel. The limit is 100 BGP routes advertised and if more than 100 the tunnel collapses causing complete failure for the TVPC in the account. 

After working with AWS engineers we identified that if we turn off route propagation we could eliminate this concern and not worry about it. This was not an issue as our workaround for the AWS route table number of routes forced us to turn off route propagation anyways.  The solution really depends on how you are using transit-vpc to route traffic in your AWS environment.
